### ISSUE TITLE
memory leak when destroy BoosterHandle* handler

### ISSUE THREAD

**Issue by cyb88** [dmlc/xgboost#1122]
Dear developers,

I found the memory that hold the model do not get released after i delete BoosterHandle.

I used a XgboostModel class to wrap all xgboost usages in my **C++ project**. there is a private BoosterHandle* handler_ in XgboostModel to do all xgboost API invoke. 

I have lots of models(nearly 2000, avg 2MB) serving in project, last week I added hot reload function in project (load new models then retire some old while my server is running ), after reload function finished I found memory usage increased from 3.1G to 6.4G, 6.4G is the sum of old and new models. so i deduce I might missed something in XgboostModel destroy function. 

**the way I initial xgboost model:**
1 handler_ = new BoosterHandle();
2 XGBoosterCreate(NULL, 0, handler_);
3 XGBoosterLoadModel(*handler_, model_path);

**the way I do predict:**
1 xgboost::DMatrix* dmat = NULL;
2 XGDMatrixCreateFromMat(feature_input, 1, feature_len, 0, (DMatrixHandle*)&dmat);
3 XGBoosterPredict(*handler_, DMatrixHandle(dmat), 0, 0, &len, &out_result);
4 delete []feature_input; delete dmat 

**the way I destroy xgboost model:**
1 XGBoosterFree(*handler_)
2 safe delete handler_ if 1 fail

this is my construct and destroy function, I put load model file in another initial function:
XgboostModel::XgboostModel(const std::string& model): 
    ModelWrapper(model) 
{
    try {
        handler_ = new BoosterHandle();
    } catch (std::exception& e) {
        CWARNING_LOG("new BoosterHandle() failed! exception: %s", e.what());
    }   
}

XgboostModel::~XgboostModel()
{
    if (handler_ != NULL) {
        int ret_tag = XGBoosterFree(*handler_);
        if (ret_tag != 0) {
            CWARNING_LOG("XGBoosterFree(*handler_) failed!");
        }   
        if (handler_ != NULL) {
            delete handler_;
            handler_ = NULL;
        }   
    }   
}

I checked the source code in XGBoosterFree(), it also do delete handler_ inside. beside this function, i didn't found other handler delete or free related API in c_api.h and c_api.cc. Is there any API to release model memory? or some Free function match XGBoosterCreate, XGBoosterLoadModel like XGBoosterFree match BoosterHandle()?

i keep debugging this problem in my project, and i did see ~XgboostModel() get executed in GDB, so i think i might miss something in real XGBOOST release.

I will be very appreciated if you can kindly give me some clues.

Thanks for your understanding and support!

**Comment by tqchen** [comment]
This should be the correct way you free the memory. I do not know exactly if memory leak happens. If it does, you might be able to check with valgrind.  It is also possible that xgboost to allocate small vectors made some small memory fragments which caused your problem.

**Comment by khotilov** [comment]
FYI: I have run the CLI and R xgboost through valgrind in CentOS 7 a couple of weeks ago, and I saw only the known "possibly lost" pseudo-leaks (see the SO link below for the explanation) whenever the first `#pragma omp` was reached first time during execution.
http://stackoverflow.com/questions/15659832/does-openmp-allocate-memory-and-free-all-after

**Comment by cyb88** [comment]
thx for kind help. I found a work around way to fix

() should be <> below, but <> will case character loose
Originally I used singleton std::map (string model_key, shared_ptr(ModelWrapper) ) in project.

when i retire model in hot reload, i tried:
1 check dict[key].use_count() for shared_ptr
2 store retired keys
3 dict.erase(key)

I GDB into 3 found ~ModelWrapper() and  XGBoosterFree() did executed, but memory use did not get releases in the end, after several time hot reload, program got killed by OS due to too much memory use

() should be <> below, but <> will case character loose
now I changed to use singleton  std::map(string model_key, ModelWrapper* )
and i delete ModelWrapper* explicitly  :
1 store retired keys
2 delete dict[key]; dict[key] = NULL
3 dict.erase(key)

the memory increased at first 2 time model reload, but keep stable in later hot reload. so i assume it is std::map memory fragment result to first memory increase. even though ModelWrapper* compared not save enough than shared_ptr<ModelWrapper>, but I can not find where i was wrong lead to memory keep increase.

thanks for all your help and suggestion! xgboost help us increased metric in daily performance!
